src/demo.rs

```diff
@@ -718,22 +752,66 @@ where
                     tracing::info!("\t{}", opt);
                     values.push(opt);
                 }
-
                 self.votes.insert(options.voteidx, values);
+
+                // Replay shunted messages if we have them. This ensures that we don't print VoteCast events for Vote we haven't seen the
+                // VoteOptions event for yet.
+                if !self.shunted_messages.is_empty() {
+                    let mut indices: Vec<usize> = vec![];
+                    for (idx, (vote_idx, shunted_msg)) in
+                        <Vec<(u32, DemoMessage)> as Clone>::clone(&self.shunted_messages)
+                            .into_iter()
+                            .enumerate()
+                    {
+                        if vote_idx == options.voteidx || self.votes.contains_key(&vote_idx) {
+                            // We only care about VoteCast event messages in the shunted messages queue.
+                            if let DemoEvent::VoteCast(_, _) = shunted_msg.event {
+                                indices.push(idx);
+                            }
+                        }
+                    }
+                    // early escape if no shunted messages to handle
+                    if indices.is_empty() {
+                        break 'voteEvent;
+                    }
+                    // Sort and reverse as we are REMOVING elements from shunted_messages, which will IMPACT every idx after.
+                    // So we start with the highest idx and move backwards
+                    indices.sort_unstable();
+                    indices.reverse();
+                    tracing::info!("Recalled {} shunted messages.", indices.len());
+                    for idx in indices {
+                        let (_, msg) = self.shunted_messages.remove(idx);
+                        // No other events will never actually be added to this shunted queue, or if they are,
+                        // they shouldn't be handled here. (And won't have been extracted here)
+                        if let DemoEvent::VoteCast(event, steamid) = msg.event {
+                            let resp = self.process_vote_cast_event(&event, &steamid, state);
+                            // These messages are guaranteed to have an output string due to the previous check for vote_idx
+                            if let Some(output) = resp {
+                                tracing::info!("{}", output);
+                            }
+                        }
+                    }
+                }
```

Comment:
I suspect a lot of this could be simplified with `Vec::retain`, something along the lines of:
```rust
self.shunted_messages.retain(|(voteidx, event, steamid)| {
    let Some(output) = self.get_vote_cast_message(event, steamid, state) else {
        return true;
    }

    tracing::info!("{output}");
    false
});
```

(If that causes a complaint about borrowing self while it's mutably borrowed, you could possibly just swap it with an empty vector temporarily while doing the operation, but hopefully that won't be necessary)
```rust
let mut temp = Vec::new();
std::mem::swap(&mut temp, &mut self.shunted_messages);
// Do your stuff here
std::mem::swap(&mut temp, &mut self.shunted_messages);
```

Reply:
Ah i hadn't thought of a retain to filter like this. Makes sense (still not inclined to thinking of closures as a spot to perform arbitrary side affecting actions (oo IO monad!!)). Needed to implement the swap because it doesn't like borrowing self immutably inside the closure for the get_vote_cast_message function call.

Reply:
> Needed to implement the swap because it doesn't like borrowing self immutably inside the closure for the get_vote_cast_message function call.

Ah well

> still not inclined to thinking of closures as a spot to perform arbitrary side affecting actions (oo IO monad!!)

Yeah I totally getcha, it does feel weird! And probably wouldn't be ideal for everything. I think in a similar case I would opt to pull the ones that need to be dealt with out with something like `Vec::extract_if` (formerly `Vec::drain_filter`), but that isn't stabilised yet so requires nightly, and the fact it changed from `drain_filter` to `extract_if` is an example of nightly code having breaking changes between versions (this change did stop a previous project of mine from compiling when I updated my compiler).